Adds a vector of default values to get_param_real_array()

adcroft opened the pull request:
The `default=` optional argument to get_param() only provides a uniform value to initialize an array of reals. This commit adds the optional `defaults=` argument that must have the same length as the `values` argument.

I've also added a few instances of this optional argument:
- by adding the `initialize_thickness_param()` procedure, selected by `THICKNESS_CONFIG = "param"`. The procedure was based on the "uniform" method, and uses the parameter `THICKNESS_INIT_VALUES` which defaults to uniform values derived from `MAXIMUM_DEPTH`
- the setting of MLD_EN_VALS in MOM_diabatic_driver.F90 which was previously using a work around to set defaults to 25, 2500, 250000 J/m2.
- two vectors of 4 values in user/user_change_diffusivity.F90

For super repos there will be some doc file changes, but no answer changes.

This PR was inspired after realizing `get_param_real_array()` could not be used in PR #737 where 16 run-time non-dimensional parameters are being added.

adcroft commented:
I see that tc2.target failed, but that's because the new parameter value isn't available with the target executable. @marshallward shouldn't we be running the target executable in the target's `.testing`?

Hallberg-NOAA commented on src/initialization/MOM_state_initialization.F90:
```diff
@@ -1011,6 +1014,69 @@ subroutine initialize_thickness_list(h, depth_tot, G, GV, US, param_file, just_r
+  is = G%isc ; ie = G%iec ; js = G%jsc ; je = G%jec ; nz = GV%ke
+
+
+  h0_def = ( G%max_depth / real(nz) ) * US%Z_to_m
```
According to the MOM6 style guide at https://github.com/NOAA-GFDL/MOM6/wiki/Code-style-guide#array-syntax, "We do not permit scalar-style expressions without the colon notation".   Please change line 1048 to read `h0_def(:) = ( G%max_depth / real(nz) ) * US%Z_to_m` to make it explicit that an array is being initialized here.

Hallberg-NOAA commented on .testing/tc2/MOM_input:
```diff
@@ -193,7 +193,7 @@ MINIMUM_DEPTH = 1.0             !   [m] default = 0.0
 MAXIMUM_DEPTH = 4500.0          !   [m]
                                 ! The maximum depth of the ocean.
-THICKNESS_CONFIG = "uniform"    !
+THICKNESS_CONFIG = "param"    !
```
If this change to the tc2 configuration had been made separately from the rest of the PR, both the code PR and the configuration change PRs would have passed the automated TC testing.  Had we done so, it would mean that the new code is not being tested by the TCs yet, but at least we would not be making exceptions to the rules about tests passing the automated testing before they are accepted.

marshallward replied:
I agree with this suggestion.  We should try to come up with some strategies to preserve testing while eliminating regression reports.

adcroft replied:
In the meantime, I think I can undo this change to tc2 if we adopt the altenative approach of making `default=` an array instead of adding `defaults=`.

adcroft replied:
I undid the modification of tc2

Hallberg-NOAA reviewed (changes requested):
This change looks like it will be a valuable new addition, but before this is accepted, I am requesting one minor change to keep this new commit in line with the overall MOM6 code style guide.

adcroft commented:
FWIW, dropping the new `defaults=` optional argument, and making `default=` a real array does work and only requires a few files to be corrected. These are the files where we use `default=` to set a vector of constant values, and all were easily changed:
```
        modified:   ALE/MOM_hybgen_regrid.F90
        modified:   core/MOM_open_boundary.F90
        modified:   initialization/MOM_state_initialization.F90
        modified:   parameterizations/vertical/MOM_diabatic_driver.F90
        modified:   tracer/oil_tracer.F90
        modified:   user/MOM_wave_interface.F90
        modified:   user/user_change_diffusivity.F90
```
I'll ask a the dev call if we prefer this version...

adcroft commented:
I tried assumed-rank arrays for `default=`, following @marshallward 's suggestion, and it looked very promising, but we discovered a feature-support problem with our favourite third compiler. :roll_eyes:  So I'm settling on the original proposal, have undone the tc2 change, and addressed the other raised issues.

adcroft commented:
Testing at https://gitlab.gfdl.noaa.gov/ogrp/mom6ci/MOM6/-/pipelines/25665
